### PR TITLE
fix: ensure head shares a chain with safe or finalized blocks

### PR DESCRIPTION
**Pull request by Rjected**
Previously, we did not check that the `head` hash in a fork choice update shared a chain with the `safe` / `finalized` fields in the same fork choice update. This check is part of the [Engine API specs](https://github.com/ethereum/execution-apis/blob/main/src/engine/paris.md#specification-1):

> Client software MUST return -38002: Invalid forkchoice state error if the payload referenced by forkchoiceState.headBlockHash is VALID and a payload referenced by either forkchoiceState.finalizedBlockHash or forkchoiceState.safeBlockHash does not belong to the chain defined by forkchoiceState.headBlockHash.

To accomplish this, two methods are added to `BlockchainTreeViewer`:
```rust
    /// Return whether or not the two block hashes share a chain. If either does not exist in the
    /// tree and are not part of the canonical chain, this should return `false`.
    fn share_chain(&self, first: &BlockHash, second: &BlockHash) -> Result<bool, Error>;

    /// Return whether or not the block is known to either the tree or is part of the canonical
    /// chain.
    fn is_block_known(&self, hash: &BlockHash) -> Result<bool, Error>;
```

Unlike other blockchain tree methods, the implementations of these methods need to access the tree externals (the DB), which is why they return a `Result`.

When rebased on #2441, this causes the `Inconsistent Head in ForkchoiceState` hive test to pass.

TODO: need a better name for `ensure_block_is_canonical` (confusing vs `BlockIndices::is_block_hash_canonical`)

### CODE REVIEW

File: crates/consensus/beacon/src/engine/mod.rs [hunk 1]
```diff
@@ -255,8 +255,51 @@ where
...
+            )))
+        }
+
+        // check if finalized is part of the same chain (and not zero)
+        if !state.finalized_block_hash.is_zero() &&
+            !self
+                .blockchain_tree
+                .share_chain(&state.head_block_hash, &state.finalized_block_hash)?
+        {
+            return Ok(OnForkChoiceUpdated::invalid_state())
```
Comment (mattsse) [inline crates/consensus/beacon/src/engine/mod.rs:293]: is this right that we can trigger the pipeline and return an invalid state?

File: crates/consensus/beacon/src/engine/mod.rs [hunk 2]
```diff
@@ -255,8 +255,51 @@ where
...
+            let target = if is_first_forkchoice &&
+                !state.safe_block_hash.is_zero() &&
+                self.get_block_number(state.safe_block_hash)?.is_none()
+            {
+                PipelineTarget::Safe
+            } else {
+                PipelineTarget::Head
+            };
+            self.forkchoice_state = Some(state);
+            // TODO: does the pipeline need to be idle here?
```
Comment (mattsse) [inline crates/consensus/beacon/src/engine/mod.rs:280]: I think this perhaps changes with #2397

but I'm not sure about this

File: crates/blockchain-tree/src/shareable.rs [hunk 3]
```diff
@@ -108,6 +108,60 @@ impl<DB: Database, C: Consensus, EF: ExecutorFactory> BlockchainTreeViewer
...
+                (None, Some(second_index)) => {
+                    // if the first is not canonical, and returned None for chain id, exit early
+                    if !tree.ensure_block_is_canonical(first)? {
+                        return Ok(false)
+                    }
+                    (second_index, first)
+                }
+                (None, None) => {
+                    // both part of canonical chain (or do not exist), exit
+                    return Ok(tree.ensure_block_is_canonical(first)? &&
```
Comment (mattsse) [inline crates/blockchain-tree/src/shareable.rs:137]: not familiar with the tree internals so no idea what's going on here, we definitely need some tests for this

**Review by mattsse — CHANGES_REQUESTED** [review]
I'd like a test for the share_chain thing

I'm a bit concerned that all the checks `on_forkchoice` are getting a bit too complex and hard to follow, but not sure how to improve

pending @rakita

**Review by rakita — CHANGES_REQUESTED** [review]
This seems overcomplicated, let us just send finalized and safe to the tree FCU function and do checks there.

Just to add, having inconsistent cached in the chain can mean that we have a bug or that CL is sending junk.

File: crates/consensus/beacon/src/engine/mod.rs (hunk 1 above)
Comment (Rjected) [reply crates/consensus/beacon/src/engine/mod.rs:293]: I'm not sure, in this case we may have `safe`, `finalized`, and `head` hashes that correspond to valid blocks in their own separate sidechains, and we may have to trigger the pipeline to retrieve or execute those blocks. When sent together as a single forkchoice update, the update is invalid if the `safe` and `head`, or `finalized` and `head`, do not belong to the same sidechain. So if we have to "set" the forkchoice state to run the pipeline, that forkchoice state could be invalid / inconsistent

File: crates/blockchain-tree/src/shareable.rs (hunk 3 above)
Comment (Rjected) [reply crates/blockchain-tree/src/shareable.rs:137]: yep, will add tests

**Comment by gakonst** [comment]
Closing as @Rjected will reimplement and given the above convo around complexity..let's revisit.
